### ISSUE TITLE
How to properly handle class imbalance in YOLOv5m

### ISSUE THREAD

**Issue by rmarkovic00** [ultralytics/yolov5#13589]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, 

I'm working on a YOLOv5m object detection project using a custom dataset with three classes: car, truck, and person. The class distribution is unbalanced (e.g. many more cars than trucks or persons).

To address this, I added the following line in my dataset.yaml file: weights: [1.0, 2.0, 3.0]
And in hyp.no-augmentation.yaml, I set: cls_pw: 1.5

My goal is to help the model better learn underrepresented classes by giving them more weight during training.

Is this the correct and recommended approach in YOLOv5 for handling class imbalance?
Or is there a better way to improve performance on rare classes?

Thank you!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rmarkovic00, thank you for your interest in YOLOv5 🚀! Please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/).

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

If this is a 🐛 Bug Report, please provide a **minimum reproducible example** to help us debug it.

## Requirements
[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```
## Environments
YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>
If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

🔎 This is an automated response. An Ultralytics engineer will review your question and provide further assistance soon!

**Comment by pderrenger** [comment]
Hi @rmarkovic00,

Yes, your approach for handling class imbalance in YOLOv5 is valid. Using class weights in dataset.yaml and adjusting cls_pw in your hyperparameters are good starting points.

Some additional strategies you could implement:

1. Data augmentation focused on minority classes (trucks and persons)
2. Targeted dataset expansion by collecting more samples of underrepresented classes
3. Adjust your IoU thresholds during evaluation to be class-specific
4. Consider using the `--rect` flag during training, which may help with detecting objects at different scales

For severe imbalance, you might also try undersampling the majority class (cars) or implementing focal loss by adjusting the hyperparameters.

YOLOv5 handles bias fairly well with proper configuration, but monitoring slice-level performance during training (how the model performs on each class separately) will help you identify if further adjustments are needed.

**Comment by ghost** [comment]
If you want to actually guide the model to better learn rare classes during training, you need to modify the loss computation or re-balance the dataset or sampling.
1. Use Focal Loss 
YOLOv5 supports Focal Loss, which is great for class imbalance

2.  Oversampling Rare Classes (Data Level)
Duplicate images with rare classes to balance the dataset.

Tools like Albumentations or custom scripts can help automate this.

3. Manual Class Weighting in the Loss Function (Advanced)
YOLOv5 doesn’t support per-class loss weighting natively during training.

But you can modify the loss function in the code (loss.py) to manually apply class weights.

Requires knowledge of PyTorch and understanding of YOLO loss structure.

4. Synthetic Data Augmentation for Rare Classes
Use tools like imgaug, Albumentations, or Roboflow to generate more samples for rare classes.

Helps increase generalization without full duplication.

**Comment by pderrenger** [comment]
Hi @soham-2006, thanks for those additional suggestions.

Building on this discussion, I'd like to add that class imbalance is a form of representation bias that can significantly impact model performance. Per our documentation on [dataset bias](https://www.ultralytics.com/blog/understanding-ai-bias-and-dataset-bias-in-vision-ai-systems), analyzing performance across slices is critical - a model with 90% overall accuracy might only achieve 60% on underrepresented classes.

@rmarkovic00 - You might also consider:

1. Using the Albumentations integration with YOLOv5 for advanced augmentations targeting minority classes - we have documentation on this at https://docs.ultralytics.com/integrations/albumentations/

2. For severe imbalance, try adjusting the `fl_gamma` parameter in your hyperparameters to increase the focus on hard examples (higher values like 1.5-2.0 can help with rare classes)

3. Monitoring class-specific metrics during training with `--verbose` flag to catch issues early

Remember that the most effective approach often combines multiple strategies rather than relying on weights alone.

**Comment by ghost** [comment]
Thanks

**Comment by pderrenger** [comment]
@rmarkovic00 - Just checking if the suggestions we've provided have been helpful for your class imbalance issue. If you've had a chance to implement any of these approaches, we'd be interested to hear about your results. Feel free to let us know if you encounter any challenges or have additional questions about implementing these strategies with your YOLOv5 model.

**Comment by n-patricia** [comment]
Hi,

About this imbalance problem, do we really need to work on duplicate images with rare classes? Isn't that yolov5 checks duplicate images and ignores these images?

Thank you.
Novi

**Comment by pderrenger** [comment]
Hi @n-patricia,

No, YOLOv5 does not automatically detect or ignore duplicate images during training. If you duplicate images containing rare classes, YOLOv5 will treat each copy as a separate training sample, which can be beneficial for addressing class imbalance.

However, be cautious with direct duplication as it can lead to overfitting on those specific images. Instead, consider using augmented versions of rare class images rather than exact duplicates, or combine oversampling with techniques like mixup or mosaic augmentation that YOLOv5 supports natively.

**Comment by n-patricia** [comment]
Hi @pderrenger,

Thank you for your response. Can we use augmentation on rare classes only? In case of using duplicate images, will parameter --patience help to prevent overfitting?

**Comment by pderrenger** [comment]
Hi @n-patricia,

Yes, you can apply augmentation selectively to rare classes only. This is actually a targeted approach that can be more effective than blanket augmentation. You can achieve this by organizing your dataset to apply different augmentation pipelines to different classes or by using custom data loading logic.

Regarding the `--patience` parameter, it helps with general overfitting by implementing early stopping when validation loss plateaus, but it's not specifically designed to address overfitting from duplicate images. For duplicates, the more effective approach is using varied augmentations on those rare class samples rather than exact copies, which naturally reduces the risk of overfitting on specific image patterns.

**Comment by n-patricia** [comment]
Hi @pderrenger,

When you said different augmentation pipelines to different classes, does it mean I need to have different hyps settings for rare classes or I can just implement certain albumentation parameters on rare classes? I see the implementation for albumentations in here utils/dataloaders.py , so I think to make adjustment for some of albumentations parameters in this section or maybe you have better approach.

<img width="797" height="830" alt="Image" src="https://github.com/user-attachments/assets/1f96e886-1879-4ed8-8cd7-2084feffadda" />

**Comment by pderrenger** [comment]
Hi @n-patricia,

For class-specific augmentation, modifying the dataloader as you've identified is one approach, but there are simpler alternatives:

1. **Pre-processing approach**: Create augmented versions of rare class images offline and add them to your dataset with different filenames - this way you maintain the standard training pipeline without code modifications.

2. **Custom dataset organization**: Structure your dataset so rare classes have multiple augmented versions in separate folders, then use standard YOLOv5 training.

3. **If modifying dataloaders.py**: You'd need to check the class label for each sample and apply different Albumentations transforms based on the class, but this requires careful implementation to avoid breaking the existing augmentation pipeline.

The pre-processing approach is often the most practical - you can use the [data augmentation techniques](https://www.ultralytics.com/blog/the-ultimate-guide-to-data-augmentation-in-2025) to create varied samples of your rare classes beforehand, which keeps your training setup simple while achieving the targeted augmentation you want.

**Comment by n-patricia** [comment]
Hi @pderrenger,

Thank you for your suggestion, will try it first and see the results.

**Comment by pderrenger** [comment]
You're welcome @n-patricia! Good luck with implementing the targeted augmentation approach - feel free to share your results or reach out if you run into any issues during implementation.

**Comment by n-patricia** [comment]
Hi @pderrenger ,

I added the augmented image to the train set. Here are the results, comparison between without (16139 images/427 backgrounds) and with the augmented (31905 images/4156 backgrounds) for 50 epochs. I use custom class weights and focal loss 1.5 for both settings.

Augmented code
```
def get_transform():
    transforms = [
        A.OneOf([
            A.GaussianBlur(blur_limit=(1,3)),
            A.RandomToneCurve(scale=0.1, p=0.5)],p=0.5),
        A.OneOf([
            A.RandomBrightnessContrast(p=1),
            A.Defocus(radius=(3,10))],
            p=0.9),
        A.OneOf([
            A.Sharpen(p=1),
            A.Blur(blur_limit=3, p=1)],
            p=0.9),
        A.OneOf([
            A.ToGray(p=1.0), A.CLAHE(clip_limit=4)], p=0.9),
        A.PlanckianJitter(
            mode='blackbody',
            temperature_limit=(3000,9000),
            sampling_method='uniform', p=0.5),
        ]
    return A.Compose(transforms)
```

**without**

<img width="1095" height="665" alt="Image" src="https://github.com/user-attachments/assets/bfda9667-9646-470a-8f50-e0707a9fa012" />

**with**

<img width="1103" height="668" alt="Image" src="https://github.com/user-attachments/assets/f746dd6e-3703-40fe-82c3-77c9c9dd116d" />

I don't see any improvement. One of the issue here is the dominant class will always appear in most of the images. For this experiment, I only use 10 classes, with 2 dominant classes. Although I filter only rare class, the two dominant classes will always be included in the augmented image. As you can see the precision is higher for both settings, do you have any approaches to detect the problem? What should I do to improve the recall and mAP?

**Comment by pderrenger** [comment]
Hi @n-patricia,

You've identified a key issue - when rare classes co-occur with dominant classes in the same images, augmenting those images still reinforces the dominant classes, which can actually hurt the balance you're trying to achieve.

A few targeted approaches to try:

1. **Crop-based augmentation**: Instead of augmenting entire images, crop regions containing only the rare classes and augment those crops as separate training samples.

2. **Monitor per-class metrics**: Use `--verbose` during training to see precision/recall for each individual class - this will show you if the rare classes are actually improving or if the dominant classes are masking the issue.

3. **Reduce augmentation intensity**: Your current augmentation pipeline is quite aggressive with multiple OneOf operations at high probabilities - try reducing the probability values or simplifying the transforms to see if you're introducing too much noise.

4. **Check actual class distribution**: Verify that your augmented dataset actually increased the representation of rare classes relative to dominant ones, not just overall image count.

The slight decrease in performance suggests the augmentations might be too aggressive or not effectively targeting the imbalance - consider starting with simpler augmentations like basic geometric transforms before adding complex photometric changes.

**Comment by n-patricia** [comment]
Hi @pderrenger,

I have data from several domains, each domain has different rare classes but the dominant classes appear in most of the domains. 
For example:
domain 1: class A, class B, class C, class D, class E, class F, class G
domain 2: class A, class B, class D, class E, class G, class H
domain 3: class A, class B, class F, class G, class H, class I
domain 4: class A, class B, class E, class G, class I, 

Just said that I had trained model from domain 1 (has 5 classes) and now I want to re-train data from domain 2 and 3 (adding 2 more classes become 7 classes) using pre-trained model from domain 1. As you can see, I have dynamic environments, where the dominant classes (A, B) appear in all domains, but the rare classes (C-I) can exist or not in all domains. 
Some questions here:
1. What is the best settings that can be used in this case? 
2. Should I train the model for each domain separately first (domain 2 and domain 3) and later combine them?
3. I am wondering would it possible to add the rare classes from different domains, just to increase the number of instances as there might some shifts from each domains, (e.g different settings when taking the samples, different lighting conditions, different image format, etc)
4. For the augmentations, should I crop the rare classes across domains, for example here I add class C, D, E, F, G from domain 1 and add into domain 2 and domain 3?

**Comment by pderrenger** [comment]
Hi @n-patricia,

For your multi-domain scenario with dynamic class distributions, here's a practical approach:

1. **Training strategy**: Train on combined domains (2+3) rather than separately - this helps the model learn shared representations for dominant classes while seeing more varied examples of rare classes across domains.

2. **Cross-domain rare class augmentation**: Yes, combining rare classes across domains is beneficial since domain shifts (lighting, settings, formats) actually help model generalization. Crop rare class instances from domain 1 and augment them into domains 2+3 training data.

3. **Incremental learning setup**: Use your domain 1 pretrained model as starting point, but freeze early layers initially and fine-tune with a lower learning rate (try 0.001 instead of default 0.01) to preserve learned dominant class features while adapting to new rare classes.

4. **Class weighting adjustment**: Since you're adding classes C, H, I that weren't in domain 1, weight these new classes higher in your dataset.yaml while keeping existing rare classes (D, E, F) moderately weighted.

The key is leveraging cross-domain diversity for rare classes while maintaining stability for dominant classes through careful learning rate and freezing strategies. Monitor per-class metrics with `--verbose` to ensure new rare classes are learning without degrading existing performance.

**Comment by n-patricia** [comment]
Hi @pderrenger ,

I have questions regarding class weighting. I saw you posted about the weight that should be added in data.yaml 
```
train: path/to/train
val: path/to/val
nc: 2
names: ['car', 'person']
class_weights: [1.0, 2.0]  # weight person class 2x higher than car
```

- Is there any differences with the weights on the hyp? Here is my settings in hyp

<img width="144" height="639" alt="Image" src="https://github.com/user-attachments/assets/8b869d0c-9985-4280-8643-4c0300f67821" />

- I use yolov5m at 1280 resolution because I have many small rare class objects that make the model difficult to learn. Is there any parameters that I need to consider to improve the results?

<img width="961" height="856" alt="Image" src="https://github.com/user-attachments/assets/73a30e1a-c77a-4e15-b2a6-11b614967972" />

- I try cross-domain rare class augmentations, and I want to make sure that would it possible to add these classes to validation set? Because there are some classes without any instances if I use the domain that I want to retrain (for example there is no class C in domain 2 and 3).

**Comment by pderrenger** [comment]
Hi @n-patricia — quick answers:

- data.yaml class_weights vs hyp: YOLOv5 does not read per‑class weights from data.yaml; hyp keys like box/cls/obj, cls_pw, obj_pw, fl_gamma are global. For true per‑class loss weighting you’d need to modify utils/loss.py or re-balance sampling/oversample minorities.

- Small rare objects at 1280: consider training an m6 model at 1280, enable --multi-scale, keep autoanchor on and, if small anchors aren’t matching well, try anchor_t: 2.0; copy_paste can help increase small-object frequency. Example hyp overrides:
  copy_paste: 0.2
  anchor_t: 2.0

- Validation: yes, include cross‑domain examples of the rare classes in val so their metrics are computed; just ensure the val set is strictly held‑out (no exact/augmented duplicates from train).

**Comment by n-patricia** [comment]
Hi @pderrenger Can I reopen this disccusion? I had yolov5m model trained with parameters as previously discussed. Just found out that yolov5m on 1280 increases FP. For example here is the confusion matrix from yolov5m and yolov5x, although both models are trained from different datasets, we tested on same test sets.

<img width="505" height="116" alt="Image" src="https://github.com/user-attachments/assets/b0df57c3-34f4-4e2e-8d90-8dbafc1a53d8" />

**Comment by pderrenger** [comment]
Hi @n-patricia, happy to keep iterating here even though the issue is closed. More FPs from the YOLOv5m model trained at 1280 are often due to the combination of higher resolution and stronger rare‑class biasing, so first make sure you’re comparing to YOLOv5x with identical inference settings (`imgsz`, `conf_thres`, `iou_thres`) and try slightly higher confidence (or class‑specific thresholds) for the dominant classes. If FPs remain high, back off the focal‑loss gamma / class weighting you added and optionally test the 1280‑trained model at a somewhat lower inference resolution (e.g. 960) to see if precision improves; for a fair comparison between m and x you’ll want both trained on the same merged dataset and evaluated with the same test configuration as shown in the [Ultralytics Docs](https://docs.ultralytics.com).
